src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs

```diff
@@ -166,6 +166,28 @@ public async Task CompilationsInCompilationReferencesIncludeGeneratedSourceFiles
             Assert.Same(compilationWithGenerator, compilationReference.Compilation);
         }
 
+        [Fact]
+        public async Task RequestingGeneratedDocumentsTwiceGivesSameInstance()
+        {
+            using var workspace = AdhocWorkspaceTests.CreateWorkspaceWithRecoverableSyntaxTreesAndWeakCompilations();
+            var analyzerReference = new TestGeneratorReference(new GenerateFileForEachAdditionalFileWithContentsCommented());
+            var project = AddEmptyProject(workspace.CurrentSolution)
+                .AddAnalyzerReference(analyzerReference)
+                .AddAdditionalDocument("Test.txt", "Hello, world!").Project;
+
+            var generatedDocumentFirstTime = Assert.Single(await project.GetSourceGeneratedDocumentsAsync());
+
+            // Fetch the compilation, and then wait for it to be GC'ed, then fetch it again. This ensures that
+            // finalizing a compilation more than once doesn't recreate things incorrectly.
+            var compilationReference = ObjectReference.CreateFromFactory(() => project.GetCompilationAsync().Result);
+            compilationReference.AssertReleased();
+            await project.GetCompilationAsync();
+
+            var generatedDocumentSecondTime = Assert.Single(await project.GetSourceGeneratedDocumentsAsync());
+
+            Assert.Same(generatedDocumentFirstTime, generatedDocumentSecondTime);
+        }
```

Comment:
in this vein, having a bunch of tests that validate getting compilations multiple times, and that that properly gets the right trees, woud be good.

Reply:
I've updated this test to also assert the tree identity is unchanged.